src/components/letter/LetterForm.tsx

```diff
@@ -0,0 +1,283 @@
+        const onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+          onAddressSelected(event.target.checked, address);
+        };
+        const key = `${address.name}:${officialAddress.officeName}`;
+        return (
+          <label className="pure-checkbox" key={key}>
+            <input type="checkbox" onChange={onChange} />
+            <>
+              <b>{address.name}</b>
+              {officialAddress.officeName && ` (${officialAddress.officeName})`}
+              , {addressToSingleLine(address)}{" "}
+              {officialAddress.link && (
+                <a target="_blank" href={officialAddress.link}>
+                  Read about their positions
+                </a>
+              )}
+            </>
+          </label>
+        );
+      })}
+    </div>
+  );
+}
+
+/** Renders all the input fields to fill in the letter and complete the transaction
+ *
+ * @return {React.ReactNode} the rendered component
+ */
+function Inputs({
+  inputs,
+  updateField,
+}: {
+  inputs: string[];
+  updateField: (key: string, value: string) => void;
+}) {
+  return (
+    <fieldset className="pure-form-aligned ">
+      {inputs.map((input) => {
+        const onChange = (
+          event: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>
+        ) => {
+          updateField(input, event.target.value);
+        };
+        return (
+          <div className="pure-control-group" key={input}>
+            <label>{_.startCase(_.toLower(input))}</label>
+            <input type="text" onChange={onChange} />
+          </div>
+        );
+      })}
+    </fieldset>
+  );
+}
+
+interface Props {
+  template: Template;
+  googleApiKey: string;
+}
+
+/** Renders the overall letter sending form
+ *
+ * @return {React.ReactNode} the rendered component
+ */
+function LetterForm({ template, googleApiKey }: Props): ReactElement {
+  const [bodyText, setBodyText] = useState(template.template);
+  const [bodyTextEdited, setBodyTextEdited] = useState(false);
+  const [myAddress, setMyAddress] = useState({} as Address);
+  const [variableMap, setVariableMap] = useState({} as Record<string, string>);
+  const [checkedAddresses, setCheckedAddresses] = useState([] as Address[]);
+  const [officials, setOfficials] = useState([] as OfficialAddress[]);
+  const [isSearching, setIsSearching] = useState(false);
+
+  // Pull all the variables out of our template
+  let variables = parseVars(template.template) || [];
+
+  // If there isn't one that includes "email", add it
+  // because we need an email to send people verification their letter went through
+  let emailKey = _.find(variables, (v) =>
+    v.toLocaleLowerCase().includes("email")
+  );
+
+  if (!emailKey) {
+    variables = [...variables, "YOUR EMAIL"];
+    emailKey = "YOUR EMAIL";
+  }
+
+  // When the inputted address changes, if it is fully specified
+  // geocode it and search for representatives' addresses
+  useEffect(() => {
+    if (
```

Comment:
factor this and the handlers into their own functions?

Reply:
they are their own functions already? I don't really see hooks passed named functions in code - cf the examples in https://reactjs.org/docs/hooks-overview.html or https://reactjs.org/docs/hooks-reference.html#useeffect

basically with hooks, you treat the enclosing function like a class and stuff lives inside it. Pulling these out into their own functions outside the body of the component function would mean passing in all the setX functions which would be a pain.